specification/sections/04-hooks.md

```diff
@@ -81,75 +87,154 @@ see: [dynamic-context paradigm](../glossary.md#dynamic-context-paradigm)
+for OpenTelemetry could be created in a `before` stage and closed in an `after` stage.
+
+```Java
+  public Optional<EvaluationContext> before(HookContext context, HookHints hints, HookData data) {
+    SpanBuilder builder = tracer.spanBuilder('sample')
+    .setParent(Context.current().with(Span.current()));
+    Span span = builder.startSpan()
+    data.set("span", span);
+  }
+
+  public void after(HookContext context, FlagEvaluationDetails details, HookHints hints, HookData data) {
+    // Only accessible by this hook for this specific evaluation.
+    Object value = data.get("span");
+    if (value instanceof Span) {
+      Span span = (Span) value;
+      span.end();
+    }
+  }
+```
 
 #### Requirement 4.3.1
 
+> `hook data` **MUST** be a structure supports definition of arbitrary properties, with keys of type `string`, and values of type `boolean | string | number | datetime | structure`.
+
+### 4.4. Hook creation and parameters
+
+#### Requirement 4.4.1
+
 > Hooks **MUST** specify at least one stage.
 
-#### Condition 4.3.2
+#### Requirement 4.4.2
+
+> `Hook data` **MUST** must be created before the first `stage` invoked in a hook for a specific evaluation and propagated between each `stage` of the hook.
+
+Example showing data between `before` and `after` stage for two different hooks.
+```mermaid
+sequenceDiagram
+actor Application
+participant Client
+participant HookA
+participant HookB
+
+Application->>Client: getBooleanValue('my-bool', myContext, false)
+activate Client
+
+Client-->>Client: create hook data for HookA
+
+Client->>HookA: before(hookContext, data: {})
+activate HookA
+
+HookA-->>HookA: data.set('key',' data for A')
+
+HookA-->>Client: (return)
+deactivate HookA
+
+Client-->>Client: create hook data for HookB
+
+Client->>HookB: before(hookContext, hints, data: {})
+activate HookB
+
+HookB-->>HookB: data.set('key', 'data for B')
+deactivate HookB
+
+Client-->>Client: Flag evaluation
+
+Client->>HookB: after(hookContext, detail, hints, data: {key: 'data for B'})
+activate HookB
+
+HookB-->>Client: (return)
+deactivate HookB
+
+Client->>HookA: after(hookContext, data: {'key': 'data for A'})
+activate HookA
+
+HookA-->>Client: (return)
+deactivate HookA
+
+Client-->>Application: true
+deactivate Client
+
+```
+
+#### Condition 4.4.2
 
 > The implementation uses the dynamic-context paradigm.
 
 see: [dynamic-context paradigm](../glossary.md#dynamic-context-paradigm)
 
-##### Conditional Requirement 4.3.2.1
+##### Conditional Requirement 4.4.2.1
 
-> The `before` stage **MUST** run before flag resolution occurs. It accepts a `hook context` (required) and `hook hints` (optional) as parameters and returns either an `evaluation context` or nothing.
+> The `before` stage **MUST** run before flag resolution occurs. It accepts a `hook context` (required), `hook hints` (optional), and `hook data` (required) as parameters and returns either an `evaluation context` or nothing.
```

Comment:
Should hook data be optional? I think most implementations could tolerate overloads for the stage functions:

```java
before(HookContext context, HookHints hints);
before(HookContext context, HookData data, HookHints hints);
```
This would mean that current hooks wouldn't break, I think.

The other option might be to add the `hook data` to the `hook context` object.

Reply:
I can think about the hook context. For the LD hooks the context was fully immutable, but for OpenFeature it isn't exactly the case being as the evaluation context within the hook context can change. So that makes it a potential home. It does make the mutability story a bit more complex.

Reply:
Updated to be in the hookContext.